cdap-data-fabric/src/test/java/co/cask/cdap/spi/metadata/MetadataStorageTest.java

```diff
@@ -877,86 +884,98 @@ public void testSearchPagination() throws IOException {
-                      .setSorting(new Sorting(ENTITY_NAME_KEY, Sorting.Order.ASC)).build(),
-                    appRecord, datasetRecord);
+    // create 10 unique random entity ids with random creation times
+    NoDupRandom random = new NoDupRandom();
+    List<MetadataEntity> entities = new ArrayList<>();
+    for (int i = 0; i < 10; i++) {
+      entities.add(MetadataEntity.ofDataset("myns", "ds" + String.valueOf(random.nextInt(1000))));
+    }
+
+    long creationTime = System.currentTimeMillis() - TimeUnit.MINUTES.toMillis(60);
+    List<MetadataRecord> records = entities.stream().map(
+      entity -> new MetadataRecord(
+        entity, new Metadata(SYSTEM, props(MetadataConstants.CREATION_TIME_KEY,
+                                           String.valueOf(creationTime + random.nextInt(1000000)),
+                                           MetadataConstants.ENTITY_NAME_KEY,
+                                           entity.getValue(entity.getType())))))
+      .collect(Collectors.toList());
+
+    // index all entities
+    mds.batch(records.stream().map(
+      record -> new MetadataMutation.Update(record.getEntity(), record.getMetadata()))
+                .collect(Collectors.toList()));
+
+    testSortedSearch(mds, records, ENTITY_NAME_KEY);
+    testSortedSearch(mds, records, CREATION_TIME_KEY);
+
+    // clean up
+    mds.batch(entities.stream().map(Drop::new).collect(Collectors.toList()));
+  }
+
+  private void testSortedSearch(MetadataStorage mds, List<MetadataRecord> records, String sortKey) throws IOException {
+    // test search by sort key itself
+    for (MetadataRecord record : records) {
+      assertResults(mds, SearchRequest.of(
+        sortKey + ":" + record.getMetadata().getProperties(SYSTEM).get(sortKey)).build(), record);
+    }
+
+    Sorting asc = new Sorting(MetadataConstants.CREATION_TIME_KEY, Sorting.Order.ASC);
+    Sorting desc = new Sorting(MetadataConstants.CREATION_TIME_KEY, Sorting.Order.DESC);
+    List<MetadataRecord> sorted = new ArrayList<>(records);
+    sorted.sort(Comparator.comparingLong(
+      r -> Long.parseLong(r.getMetadata().getProperties().get(new ScopedName(SYSTEM, CREATION_TIME_KEY)))));
+    List<MetadataRecord> reverse = new ArrayList<>(sorted);
+    Collections.reverse(reverse);
+
+    testSortedSearch(mds, asc, sorted);
+    testSortedSearch(mds, desc, reverse);
+  }
+
+  private void testSortedSearch(MetadataStorage mds, Sorting sorting, List<MetadataRecord> sorted) throws IOException {
+    // search once with sort by creation time and validate that records are in order
+    SearchResponse response = mds.search(SearchRequest.of("*").setSorting(sorting).build());
+    Assert.assertEquals(sorted, response.getResults());
+
+    // search for a window of results
+    assertInOrder(mds, SearchRequest.of("*").setSorting(sorting).setOffset(0).setLimit(4).build(),
+                  sorted.get(0), sorted.get(1), sorted.get(2), sorted.get(3));
```

Comment:
Can use sorted.sublist(0, 4) to reduce verbosity here and below

Reply:
if the method assertInOrder did not expect an array (varargs)

Reply:
ok, I introduced a variation that takes a list.